Add functional collective all_to_all_single and support it in Inductor

yf225 opened the pull request:
Summary: `all_to_all_single` is being used by all the important All2All workloads (e.g. alltoall_pooled / alltoall_sequence / alltoallv). This PR adds functional collective `all_to_all_single` and support it in Inductor, to unlock our exploration on internal workloads.

Differential Revision: D48013416



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel @anijain2305

kumpera commented on torch/distributed/_functional_collectives.py:
```diff
@@ -291,6 +291,32 @@ def reduce_scatter_tensor_coalesced(
 
+def all_to_all_single(
+    self: torch.Tensor,
+    output_split_sizes: List[int],
```
We cannot use object list for input/output split sizes, they must be tensors or dynamo will incorrectly bake them as constants.

wanchaol replied:
I was thinking that we should probably follow the xla semantic of alltoall https://www.tensorflow.org/xla/operation_semantics#alltoall, that would make dynamo and inductor be much happier. The output_split_sizes and input_split_sizes would require pass in constant list (even if we make them as tensor, inductor technically can't do anything better rather than be blind on it).

kumpera commented on torch/distributed/_functional_collectives.py:
```diff
@@ -291,6 +291,32 @@ def reduce_scatter_tensor_coalesced(
     return list(map(_maybe_wrap_tensor, tensor_list))
 
 
+def all_to_all_single(
```
We should follow the design of other collectives and make the concat dim an explicit argument.

ezyang replied:
@kumpera basically, adding these lines, is that right? https://github.com/pytorch/pytorch/blob/f4b464c6d0c090c4a29c4148ea289812f2720fea/torch/distributed/_functional_collectives.py#L175-L176

yf225 replied:
The current design is actually sufficient for current use cases (and matches existing `dist.all_to_all_single` API). Maybe we can make adding concat dim a separate work item? (which would make a nice onboarding task too)

kumpera replied:
We can add it later.
This is an old ask by @wanchaol as shuffling around collectives is a common source of bad overlap.

kumpera commented on torch/distributed/_functional_collectives.py:
```diff
@@ -291,6 +291,32 @@ def reduce_scatter_tensor_coalesced(
+def all_to_all_single(
+    self: torch.Tensor,
+    output_split_sizes: List[int],
+    input_split_sizes: List[int],
```
both lists should be optional in case the use wants exact splitting.

kumpera commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -263,3 +263,26 @@ def _reduce_scatter_tensor_coalesced_fallback(output_tensors, input_tensors, op,
+
+    out_size = list(input.size())
+    out_size[0] = sum(output_split_sizes)
+    out_tensor = input.new_empty(out_size)
```
This only works with 1D tensors, either fix that or assert on dim > 1

yf225 replied:
@kumpera I thought about this some more and believe `input.dim() == 1` (1d input) also works, I added an assert and a test case for this

kumpera commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -263,3 +263,26 @@ def _reduce_scatter_tensor_coalesced_fallback(output_tensors, input_tensors, op,
+    out_size[0] = sum(output_split_sizes)
+    out_tensor = input.new_empty(out_size)
+
+    work = c10d.all_to_all_single(
```
gloo doesn't support all_to_all_single, you should use all_to_all in that case.

yf225 replied:
thanks - checked and it seems gloo supports all_to_all_single now, I added a CPU test to ensure

kumpera commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -214,6 +214,41 @@ def compile(func, example_inputs):
+    @skip_if_lt_x_gpu(2)
+    # TODO: somehow inductor bg compile threads are causing hangs at exit with distributed work dtor
+    @patch.object(torch._inductor.config, "compile_threads", 1)
+    def test_all_to_all_single_inductor(self):
```
Please add a test for alll_to_all using non-constant split sizes.

wconstab commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -33,14 +33,22 @@
+            node.get_device(),
+            node.get_dtype(),
+            V.graph.sizevars.simplify(sympy_product(size)),
+            # Detect gaps in tensor storage caused by strides
```
i don't follow what this does.  you multiply out (s2-1 * stride2, s1-1 * stride1...) and what does that give you? and why is it useful to return that quantity instead of the actual stride?

yf225 replied:
I didn't add this part - was copied from previous code

ezyang replied:
The point is that the last element size hint says how big the extent of the tensor is, even if pieces inside it aren't actually used (because striding skips over it).

Suppose you have a 2x2 tensor with stride 4x1, versus 1x4. In both cases, the extent of the tensor is (4*1 + 1*1 == 5), so you can represent both variants in the same amount of memory, so you can reuse the buffer in this case.

Feel free to write this comment into the code.

wconstab commented on torch/_inductor/ir.py:
```diff
@@ -5261,3 +5262,97 @@ def codegen_collective(self, wrapper, output_name, input_names):
+            new_size[0] = output_split_sizes_sum_s
+            # FlexibleLayout ctor used below needs to know which sizevar in output size
+            # is okay to skip the `isinstance(s, (Expr, int))` check for.
+            V.graph.data_dep_symbol_to_expr_str[str(output_split_sizes_sum_s)] = None
```
ok so, is it correct that you use this dict as a 3 step pipeline?
1) put a key = none to signal that key needs to be populated
2) in codegen, update the key to an expression which..i guess you couldn't have created at step 1 because... idk, maybe the output names aren't known yet?
3) in wrapper codegen, utilize the expression from (2) to create the sizevar that will be used to alloc the output that is referenced by codegen in (2)?

it's pretty convoluted but i can see that it would work.  I just wonder if there would be a way to make it cleaner (perhaps in a later PR)?

(oh, and it'd be good to put a comment somewhere that explains this 3 steps)

yf225 replied:
thanks - after a few iterations, no more need for this 3-step pipeline (yay!)

wconstab commented on torch/_inductor/lowering.py:
```diff
@@ -4610,6 +4610,12 @@ def reduce_scatter_tensor_coalesced(self, reduceOp, tag, ranks, group_size):
         return list(map(TensorBox.create, result))
 
+    @register_lowering(c10d_functional.all_to_all_single)
+    def all_to_all_single(self, output_split_sizes, input_split_sizes, tag, ranks, group_size):
```
I thought @eellison had a trick for getting the meta-inferred output size here.  I guess it didn't pan out for this case but I was still curious to learn about it regardless.  And is the reason that you didn't go that route because you need the symbolic expression not the constant?

yf225 replied:
Elias taught me how to get the meta-inferred output size - I used it within the `ir.AllToAllSingle.create` method: `output_split_sizes_sum_s = V.graph.current_node.meta['val'].size()[0]`

yf225 replied:
actually no more need to pass down meta-inferred output size after switching to the new design using ExternKernelAlloc (which also makes it much cleaner)

wconstab reviewed (commented):
i mostly reviewed the inductor lowerings part, and i think that looks correct, although i'd hope we can avoid the global var scheme somehow.  Deferring to @kumpera on the the collective op signature.

yf225 commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -214,6 +232,126 @@ def compile(func, example_inputs):
+                ranks=ranks,
+                group_size=group_size,
+            )
+            a2a = torch.ops.c10d_functional.wait_tensor(a2a)
```
Note to self: add `out = a2a / a2a.sum(dim=0)` to test persistent reduction after a2a, to see whether Inductor can reason about the symbolic sizevar in output shape

ezyang commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -8,7 +8,8 @@ def signature_of(arg, *, size_dtype: str):
 
     if isinstance(arg, TensorArg):
-        tye = JITFunction._type_of(arg.dtype)
+        arg_dtype = V.graph.name_to_coerced_dtype.get(arg.buffer, arg.dtype)
```
What's going on here?

yf225 replied:
we no longer need this

ezyang commented on torch/_inductor/graph.py:
```diff
@@ -194,6 +194,8 @@ def __init__(
         self.name_to_buffer: Dict[str, ir.ComputedBuffer] = {}
         self.name_to_users: DefaultDict[str, List[ir.IRNode]] = defaultdict(list)
+        self.name_to_coerced_dtype: Dict[str, torch.dtype] = {}
+        self.data_dep_symbol_to_expr_str: Dict[str, str] = {}
```
Some docs for these would be appreciated

yf225 replied:
removed the need for `data_dep_symbol_to_expr_str`

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -33,14 +33,22 @@
+            node.get_device(),
+            node.get_dtype(),
+            str(size),
+            str(stride),
```
@jansel says to use `sympy_str` here instead of str, it will be faster. Also, recording the entire stride is not necessary, you should use the last element calculation. Actually, I would suggest NOT having the top level conditional here; instead, you want to add a `handle_unbacked` call around the fourth tuple element, which tests if last_element is size hint'able, and if it isn't switches to `sympy_str`. (Actually, it's kind of weird that line 40 can take sympy symbols, I wonder what's up with that...)

yf225 replied:
Found that this is no longer needed after switching to ExternKernelAlloc

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -212,6 +220,12 @@ def plan(self, state: MemoryPlanningState):
     def codegen(self, code: IndentedBuffer):
         assert self.node.get_name() not in V.graph.removed_buffers
+        for s in self.node.get_size():
+            s_str = str(s)
```
You don't have to stringify; instead, test if you have a Symbol and that Symbol is data dependent, and if so make the binding.

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -212,6 +220,12 @@ def plan(self, state: MemoryPlanningState):
+            if s_str in V.graph.data_dep_symbol_to_expr_str:
+                expr_str = V.graph.data_dep_symbol_to_expr_str[s_str]
+                assert expr_str is not None
+                code.writeline(f"{s_str} = {expr_str}")
```
I'm a little unsure whether or not this is the correct place to put this logic. When I was hacking through, I placed equivalent code in https://gist.github.com/ezyang/1293a41299604c44310341b7540eabcb#file-gistfile0-txt-L108-L121 What made you decide to put it here?

ezyang replied:
cc @jansel for thoughts

yf225 replied:
currently the generated code is roughly:
```python
i3 = buf0.sum().item()     # <----- this is the line of concern
buf2 = empty_strided((i3, s1), (s1, 1), device='cuda', dtype=torch.float32)
buf3 = [buf2]
buf3_work = dist.all_to_all_single(output=buf3[0], ...)
```
`buf2` which uses `i3` as dim size is input to `all_to_all_single`, so we might need to generate `i3` computation line here instead of in `MultiOutput`

yf225 replied:
Removed the need for `data_dep_symbol_to_expr_str`

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -602,6 +616,9 @@ def append_precomputed_sizes_to_prefix(self):
     def codegen_python_sizevar(self, x: Expr) -> str:
+        # If sizevar is an expected data-dependent symbol, then just use its string repr.
+        if str(x) in V.graph.data_dep_symbol_to_expr_str:
+            return pexpr(x)
```
Nyeh, this shouldn't be necessary. How does simplify choke on data-dependent symbol?

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -2808,8 +2808,9 @@ def require_stride1(cls, x):
     def require_stride_order(cls, x, order):
-        if x.get_numel() == 0:  # Layout doesn't matter
-            return x
+        if not any([str(s) in V.graph.data_dep_symbol_to_expr_str for s in x.get_size()]):
```
Actually, for data dependent symbols, we will typically (unsoundly) assume that they are nonzero, so conventional reasoning using SymInt value ranges would be able to infer if it's zero in the test here. In your AllToAllSingle, you probably failed to `constrain_range` appropriately; check nonzero for an example constrain range.

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5261,3 +5262,97 @@ def codegen_collective(self, wrapper, output_name, input_names):
+
+
+class AllToAllSingle(OutOfPlaceCollectiveKernel):
+    count = itertools.count()
```
unused?

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5261,3 +5262,97 @@ def codegen_collective(self, wrapper, output_name, input_names):
+    def codegen_collective(self, wrapper, output_name, input_names):
+        # When output_split_sizes is specified, output shape of all_to_all_single is
+        # data-dependent on output_split_sizes, so we store the shape computation formula
+        # and during codegen time write this shape computation line before output buffer allocation.
```
I'm not really a fan of organizing the code this way. Too much action at a distance. I think I still prefer for the meta function to perform the allocation, and then we automatically insert write outs for all unbacked ints the first time they come into scope. (This would also be a good place to add runtime asserts, which we need to add soon.)

yf225 replied:
Removed the need for `data_dep_symbol_to_expr_str`

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5261,3 +5262,97 @@ def codegen_collective(self, wrapper, output_name, input_names):
+            if inp is None:
+                input_strs.append("None")
+            else:
+                input_strs.append(f"{output_name}_inputs[{self.original_inputs.index(inp)}].tolist()")
```
What's going on here?

ezyang replied:
(For efficiency reasons, it's much better if we don't call tolist here and let the kernel manage it.)

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5261,3 +5262,97 @@ def codegen_collective(self, wrapper, output_name, input_names):
+            output_split_sizes_realized = cls.realize_input(output_split_sizes)
+            # NCCL/GLOO alltoall_base also assumes output_split_sizes to be int64,
+            # so hardcoding int64 here should be okay.
+            V.graph.name_to_coerced_dtype[output_split_sizes_realized.data.name] = torch.int64
```
Yeah looking at the setter site, I also still don't understand what this means

yf225 replied:
we no longer need this

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1784,7 +1785,7 @@ def codegen_kernel(self, name=None):
 
         size_hints = [
-            next_power_of_2(V.graph.sizevars.size_hint(numel)) for numel in self.numels
+            next_power_of_2(V.graph.sizevars.size_hint(numel)) if not free_symbols(numel) else 64 for numel in self.numels
```
how did you choose 64 here? 

We should still be using the size_hint of the numel if it exists. 

We could also check if the symbols in the numel have bounds on them, and do something with that.

yf225 replied:
thanks! just realized it's a relic and we don't need this hardcoded hint anymore.

yf225 replied:
discussed with Elias offline, and added `shape_env.var_to_size_hint` to handle this

ezyang commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -304,3 +304,32 @@ def _reduce_scatter_tensor_coalesced_fallback(output_tensors, input_tensors, op,
+    group = c10d._find_or_create_pg_by_ranks_and_tag(tag, ranks, group_size)
+
+    if isinstance(input_split_sizes, torch.Tensor):
+        input_split_sizes = input_split_sizes.tolist()
```
Hmm. Is there any particular reason you call tolist here, as opposed to just taking a list of ints as arguments (you call tolist outside of the collective). Is it because the list could be very long?

yf225 replied:
I believe Dynamo will hardcode the `List[int]` input arguments as constants instead of treating them as dynamic inputs like tensors, and it's why I make `input_split_sizes` and `output_split_sizes` tensors instead

ezyang replied:
If it's `SymInt[]` it wouldn't be hardcoded. If you're actually extracting out the tensor values to ints at runtime it would be better to use this.

yf225 replied:
discussed with Ed offline - will use tensor as input type instead, because list-type input to `output_split_sizes` doesn’t have a special data structure like KJT that we can insert conversion logic for converting ints to unbacked symints.

ezyang commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -23,6 +24,23 @@
+    capture_scalar_outputs_prev_value = torch._dynamo.config.capture_scalar_outputs
+    torch._dynamo.config.dynamic_shapes = True
+    torch._dynamo.config.capture_dynamic_output_shape_ops = True
+    torch._dynamo.config.capture_scalar_outputs = True
```
you don't need to do this by hand; we have torch._dynamo.config.patch(...) which can do this for you

ezyang commented on torch/_inductor/graph.py:
```diff
@@ -204,6 +204,7 @@ def __init__(
         self.unaligned_buffers: Set[str] = set()
         self.name_to_buffer: Dict[str, ir.ComputedBuffer] = {}
         self.name_to_users: DefaultDict[str, List[ir.IRNode]] = defaultdict(list)
+        self.name_to_coerced_dtype: Dict[str, torch.dtype] = {}
```
I'm still not convinced by this

yf225 replied:
we no longer need this

ezyang commented on torch/_inductor/lowering.py:
```diff
@@ -734,7 +734,10 @@ def expand(x, sizes):
-    if x_size_product > 0:
+    if (
+        not any(V.graph.sizevars.shape_env.is_unbacked_symint(s) for s in sizes)
+        and x_size_product > 0
```
To be clear, is it the `x_size_product > 0` that is failing? The condition here seems flipped; if I can't tell if `x_size_product` is non-zero or not, it seems better to assume that it is non-zero, because that's more common.

yf225 replied:
Found that we no longer need this change after switching to use ExternKernelAlloc

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5567,3 +5571,101 @@ def codegen_collective(self, wrapper, output_name, input_names):
+    ):
+        super().__init__(layout, inputs, constant_args)
+        self.inputs = inputs
+        self.inputs_allow_none = inputs_allow_none
```
These two fields are kind of weird, not really precedent for it. (Maybe it's OK, I don't have an intuitive sense in IR for if it is ok or not)

yf225 replied:
Updated to use `arg_index_to_input_index` dict instead, to hopefully make it more clear for posterity

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5567,3 +5571,101 @@ def codegen_collective(self, wrapper, output_name, input_names):
+        wrapper.add_import_once("import torch.distributed.distributed_c10d as c10d")
+        wrapper.add_import_once(
+            "import torch.distributed._functional_collectives_impl as fun_col_impl"
+        )
```
If you had used OutOfPlaceCollective kernel this would have been taken care of for you, no?

ezyang replied:
In fact it seems you've straight copy pasted it here. Seems like you should have used OutOfPlaceCollectiveKernel?

yf225 replied:
OutOfPlaceCollectiveKernel inherits from `ExternKernel` instead of `ExternKernelAlloc` that we need - I updated the PR to add `CollectiveKernelAlloc` (subclass of `ExternKernelAlloc`) and have `AllToAllSingle` inherit from it, and `CollectiveKernelAlloc` and `CollectiveKernel` will have shared common code.

ezyang commented:
At a high level this looks good. There's some details about precisely fitting this into the prevailing codebase (which I am not the best to answer, thats @kumpera), but it seems like only modest adjustments should be needed.

yf225 committed:
Add functional collective all_to_all_single and support it in Inductor (#106655)

Summary:
`all_to_all_single` is being used by all the important All2All workloads (e.g. alltoall_pooled / alltoall_sequence / alltoallv). This PR adds functional collective `all_to_all_single` and support it in Inductor, to unlock our exploration on internal workloads.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel chenyang78 kadeng muchulee8 aakhundov anijain2305

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106655

Differential Revision: D48013416

Pulled By: yf225

fbshipit-source-id: b6d074094bbeac834433cb09707aff9c0b7280c7

yf225 commented on torch/_inductor/ir.py:
```diff
@@ -5567,3 +5571,101 @@ def codegen_collective(self, wrapper, output_name, input_names):
+"""
+
+
+class AllToAllSingle(ExternKernelAlloc):
```
added `CollectiveKernelAlloc` (subclass of `ExternKernelAlloc`) and had `AllToAllSingle` inherit from it, and try to achieve more code sharing between `CollectiveKernelAlloc` and existing `CollectiveKernel`.

yf225 replied:
updated: changed to inherit from OutOfPlaceCollectiveKernel instead

pytorchbot committed:
Change to use SymInt[]

jansel commented on torch/_inductor/ir.py:
```diff
@@ -5567,3 +5593,89 @@ def codegen_collective(self, wrapper, output_name, input_names):
+                arg_index_to_input_index[i] = None
+
+        return AllToAllSingle(
+            layout=inputs[0].get_layout(),
```
Shouldn't this be the layout of the output?  I thought it needed data-dependent shapes?  This looks simpler than I expected.

yf225 replied:
thanks - yes it's my oversight, I switched back to constructing output size with unbacked symint as first dim

pytorchbot committed:
minor rename

pytorchbot committed:
todo comment

pytorchbot committed:
fix layout issue

wconstab commented on torch/_inductor/ir.py:
```diff
@@ -5567,3 +5593,71 @@ def codegen_collective(self, wrapper, output_name, input_names):
+    because its output shape is data-dependent on input `output_split_sizes`.
+    Instead of letting Inductor manage allocation of output buffer that has
+    data-dependent dim size which is difficult to do, we resort to using
+    `ExternKernelAlloc` and letting the underlying `fun_col_impl._all_to_all_single`
```
i'm confused how this is different from other out-of-place collectives? for those, inductor also does not alloc and relies on c10d to perform the allocation.  the 'inplace' ones are special in that they hijack an input buffer created by inductor and let that be used in place.

wconstab replied:
we seem to have created a chimera in 'CollectiveKernel' since like ExternKernelAlloc, it implements 'should_allocate' to return False, but unlike ExternKernelAlloc, it will fail certain isinstance checks scattered through inductor.  This probably isn't good.  Did you hit one of those cases and hence needed to subclass ExternKernelAlloc?  Im wondering if we should have made 'OutOfPlaceCollectiveKernel' subclass from ExternKernelAlloc, but didn't since that opens the door to multiple-inheritance problems.  cc @kumpera

yf225 replied:
thanks - moved back to subclass from `OutOfPlaceCollectiveKernel`. Found mainly 3 differences compared to other existing out-of-place collectives:
1. `create_output_buffers` is not called, because we rely on the underlying `fun_col_impl._all_to_all_single` kernel to do the output buffer allocation.
2. `fun_col_impl._register_tensor_work` is not emitted, because the underlying `fun_col_impl._all_to_all_single` kernel already calls it.
3. `c10d._find_or_create_pg_by_ranks_and_tag` is not emitted, because the underlying `fun_col_impl._all_to_all_single` kernel already calls it.

(documented in code as well)

pytorchbot committed:
Revert "fix layout issue"

This reverts commit b5119504b56259e8c3ecffe35ebc6685ceedcb0a.

pytorchbot committed:
Revert "todo comment"

This reverts commit dc8237c8359fda1ec58f8b2999dec52c43d10464.

pytorchbot committed:
Revert "minor rename"

This reverts commit 47f0365ab6d5e66dd890249df0ea648a4841e06b.

pytorchbot committed:
Revert "Change to use SymInt[]"

This reverts commit 589d7e61517f904a4bb671f94848024433119804.

pytorchbot committed:
Revert "Revert "todo comment""

This reverts commit 15bd3b9f39779852a6c3b12b6f34bff4a5b8d039.

pytorchbot committed:
Switch back to using tensor input

pytorchbot committed:
clean up unused

pytorchbot committed:
clean up unused

pytorchbot committed:
clean up and add code check in test

pytorchbot committed:
add size hint for unbacked symint

pytorchbot committed:
minor comment

pytorchbot committed:
add test case for 1d input

pytorchbot committed:
added comment for buffer_reuse_key

pytorchbot committed:
use sympy_str in buffer_reuse_key

pytorchbot committed:
added comment for symbolic_hint

pytorchbot committed:
updated comment

pytorchbot committed:
updated comment

yf225 commented on torch/_inductor/sizevars.py:
```diff
@@ -366,6 +366,9 @@ def symbolic_hint(self, expr: Expr) -> Expr:
             return int(expr)
+        # Replace unbacked symints with their size hints if exists
+        if any(self.shape_env.is_unbacked_symint(s) for s in free_symbols):
+            expr = sympy_subs(expr, self.shape_env.var_to_size_hint)
```
talked to Elias and Horace about this:

Inductor guarantees that the generated code remains correct even if the size hint doesn't exactly match the actual size, as size hint is just the best attempt at guessing the actual size. The closer the guess is, the more performant the generated code will be. However, if we don't resolve the unbacked symint to the concrete size hint here, we will immediately run into error when we try to convert the result of this function to int type.

(added in comment as well)

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2009,6 +2009,8 @@ def __init__(
         # Maps symbolic ints to their original concrete values
         # Currently populated from tensors
         self.var_to_val: Dict[sympy.Symbol, sympy.Integer] = {}
+        # Maps symbolic ints to their size hints. Only used for unbacked symints.
```
If it's only used for unbacked symint, the name should probably reflect it. Also, you seem to be accessing this field directly from outside of symbolic_shapes; mediate it through a method or something please.

ezyang replied:
(I'm expecting some mechanism to populate this by user annotation eventually. Also, it would be good to explain in the comment here what this is being used for.)

ezyang commented on torch/_inductor/codegen/triton.py:
```diff
@@ -12,6 +12,7 @@
 
 import torch
 
+from torch.fx.experimental.symbolic_shapes import free_symbols
```
dead

ezyang commented on torch/_inductor/sizevars.py:
```diff
@@ -366,6 +366,16 @@ def symbolic_hint(self, expr: Expr) -> Expr:
+        # we will immediately run into error when we try to convert the result of this function
+        # to int type.
+        if any(self.shape_env.is_unbacked_symint(s) for s in free_symbols):
+            expr = sympy_subs(expr, self.shape_env.var_to_size_hint)
```
My main question about this is whether or not all use sites of `var_to_size_hint` are OK with this kind of hint on the unbacked SymInt. `symbolic_hint` is invoked by `size_hint` which has many sites. Are we sure these are all OK?

ezyang replied:
It would be good to have a much beefier comment on `size_hint` saying exactly what this does, esp in the case of unbacked SymInt.

ezyang replied:
Also, pursuant to https://github.com/pytorch/pytorch/pull/106655/files#r1316332524 this will be useful for direct ShapeEnv users (non-Inductor). For example, a good size hint will help the partitioner make a good decision.

ezyang commented on torch/_inductor/sizevars.py:
```diff
@@ -366,6 +366,16 @@ def symbolic_hint(self, expr: Expr) -> Expr:
         free_symbols = expr.free_symbols
```
The comment above

```
        # Substitute all hints into expr, but leave unbacked symints alone
```

is now out of date

ezyang commented on torch/distributed/_functional_collectives.py:
```diff
@@ -495,6 +528,18 @@ def mk_out_tensor(input):
+        ctx = get_ctx()
+        # `output.shape[0]` is `sum(output_split_sizes)`
+        # which is data-dependent, so we use symint to represent it here.
+        output_shape_first_dim = ctx.create_unbacked_symint()
```
You probably want `constrain_range_as_size` here or something

ezyang commented on torch/_inductor/lowering.py:
```diff
@@ -734,7 +734,7 @@ def expand(x, sizes):
 
     x_size_product = V.graph.sizevars.size_hint(sympy_product(x.get_size()))
-    if x_size_product > 0:
+    if x_size_product > 0 and not any([V.graph.sizevars.shape_env.is_unbacked_symint(s) for s in sizes]):
```
The heuristic here is confusing to me. If we think the input hints to be non-empty, why would we SUPPRESS realizing the input if some of them are unbacked? It seems to me we want to realize it.

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5347,6 +5355,9 @@ def codegen(self, wrapper):
         )
+        for i, s in enumerate(self.get_size()):
+            if V.graph.sizevars.shape_env.is_unbacked_symint(s):
+                wrapper.writeline(f"{s} = {self.get_name()}.size({i})")
```
I'm not sure this is exactly the right spot to put this codegen, but we can move it later if needed

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5567,3 +5578,117 @@ def codegen_collective(self, wrapper, output_name, input_names):
+            output_shape_first_dim_s = V.graph.current_node.meta['val'].size()[0].node.expr
+            new_size[0] = output_shape_first_dim_s
+        # In the average case, `output.shape[0]` equals to `input.shape[0]`, so we use that as size hint
+        V.graph.sizevars.shape_env.var_to_size_hint[output_shape_first_dim_s] = x_realized.get_size()[0]
```
This should also be calling some method on ShapeEnv

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -5567,3 +5578,117 @@ def codegen_collective(self, wrapper, output_name, input_names):
+        new_size = x_realized.get_size()
+        output_shape_first_dim_s = None
+        if output_split_sizes is not None:
+            output_shape_first_dim_s = V.graph.current_node.meta['val'].size()[0].node.expr
```
I feel it may be possible for this size to be an int; e.g., if a user asserts the unbacked symint == 5, then we should have refined it by this point.

ezyang reviewed (approved):
Just small stuff. Good work.

pytorchbot committed:
try List[int]

pytorchbot committed:
try to switch to SymInt[]

pytorchbot committed:
clean up

pytorchbot committed:
replace specialized 0/1 with unbacked symint with range

pytorchbot committed:
clean up

pytorchbot committed:
clean up

pytorchbot committed:
clean up

pytorchbot committed:
clean up

pytorchbot committed:
WIP replace with unbacked symint in Dynamo

pytorchbot committed:
ir.DynamicScalar codegen

pytorchbot committed:
clean up

pytorchbot committed:
clean up

pytorchbot committed:
clean up

pytorchbot committed:
clean
